Allow use of Ctrl+Space keybinding on Mac OS X - bug 89069.

Index: /Users/toddw/as/komodo-devel/mozilla/build/moz191-ko6.1/mozilla/widget/src/cocoa/nsChildView.mm
--- widget/src/cocoa/nsChildView.mm	Thu Jul 01 01:37:30 2010 -0700
+++ widget/src/cocoa/nsChildView.mm	Thu Feb 10 11:16:05 2011 -0800
@@ -4862,8 +4868,14 @@
       outGeckoEvent->charCode = [chars characterAtIndex:0];
     
     // convert control-modified charCode to raw charCode (with appropriate case)
-    if (outGeckoEvent->isControl && outGeckoEvent->charCode <= 26)
-      outGeckoEvent->charCode += (outGeckoEvent->isShift) ? ('A' - 1) : ('a' - 1);
+    if (outGeckoEvent->isControl && outGeckoEvent->charCode <= 26) {
+      /* XXX ACTIVESTATE: charCode of ctrl-space is 0x00 - bug 89069. */
+      if (outGeckoEvent->charCode == 0) {
+        outGeckoEvent->charCode = 0x20;
+      } else {
+        outGeckoEvent->charCode += (outGeckoEvent->isShift) ? ('A' - 1) : ('a' - 1);
+      }
+    }
 
     // Accel and access key handling needs to know the characters that this
     // key produces with Shift up or down.  So, provide this information
@@ -5723,15 +5735,25 @@
     unsigned int modifierFlags =
       nsCocoaUtils::GetCocoaEventModifierFlags(theEvent) & NSDeviceIndependentModifierFlagsMask;
     if (modifierFlags == NSControlKeyMask && [[theEvent charactersIgnoringModifiers] isEqualToString:@" "]) {
-      nsMouseEvent contextMenuEvent(PR_TRUE, NS_CONTEXTMENU, [self widget], nsMouseEvent::eReal, nsMouseEvent::eContextMenuKey);
-      contextMenuEvent.isShift = contextMenuEvent.isControl = contextMenuEvent.isAlt = contextMenuEvent.isMeta = PR_FALSE;
-      PRBool cmEventHandled = mGeckoChild->DispatchWindowEvent(contextMenuEvent);
-      [self maybeInitContextMenuTracking];
-      // Bail, there is nothing else to do here.
-      PRBool handled = (cmEventHandled || mKeyDownHandled);
-      mCurKeyEvent = nil;
-      mKeyDownHandled = PR_FALSE;
-      return handled;
+
+      /* XXX ACTIVESTATE: allowing to customize ctrl+space keybinding - bug 89069. */
+      nsCOMPtr<nsIPrefBranch> prefs = do_GetService(NS_PREFSERVICE_CONTRACTID);
+      PRBool useCtrlSpaceContextMenu = PR_FALSE;
+      if (prefs) {
+        nsresult rv = prefs->GetBoolPref("ui.use_ctrl_space_contextmenu", &useCtrlSpaceContextMenu);
+      }
+
+      if (useCtrlSpaceContextMenu) {
+        nsMouseEvent contextMenuEvent(PR_TRUE, NS_CONTEXTMENU, [self widget], nsMouseEvent::eReal, nsMouseEvent::eContextMenuKey);
+        contextMenuEvent.isShift = contextMenuEvent.isControl = contextMenuEvent.isAlt = contextMenuEvent.isMeta = PR_FALSE;
+        PRBool cmEventHandled = mGeckoChild->DispatchWindowEvent(contextMenuEvent);
+        [self maybeInitContextMenuTracking];
+        // Bail, there is nothing else to do here.
+        PRBool handled = (cmEventHandled || mKeyDownHandled);
+        mCurKeyEvent = nil;
+        mKeyDownHandled = PR_FALSE;
+        return handled;
+      }
     }
 
     nsKeyEvent geckoEvent(PR_TRUE, NS_KEY_PRESS, nsnull);
